k2/csrc/connect.cu

```diff
@@ -0,0 +1,497 @@
+    const int32_t *fsas_row_splits2_data = fsas_.RowSplits(2).Data(),
+                  *states_data = cur_states.values.Data();
+    char *accessible_data = accessible_.Data();
+    K2_EVAL(
+        c_, cur_states.NumElements(), lambda_set_arcs_and_accessible_per_state,
+        (int32_t states_idx01)->void {
+          int32_t fsas_idx01 = states_data[states_idx01],
+                  num_arcs = fsas_row_splits2_data[fsas_idx01 + 1] -
+                             fsas_row_splits2_data[fsas_idx01];
+          num_arcs_per_state_data[states_idx01] = num_arcs;
+          // Accessible
+          accessible_data[fsas_idx01] = 1;
+        });
+    ExclusiveSum(num_arcs_per_state, &num_arcs_per_state);
+
+    // arcs_shape `[fsas][states cur][arcs]
+    RaggedShape arcs_shape = ComposeRaggedShapes(
+        cur_states.shape, RaggedShape2(&num_arcs_per_state, nullptr, -1));
+
+    // Each arc that generates a new state (i.e. for which
+    // arc_renumbering.Keep[i] == true) will write the state-id to here (as an
+    // idx01 into fsas_).  Other elements will be undefined.
+    Array1<int32_t> next_iter_states(c_, arcs_shape.NumElements());
+
+    // We'll be figuring out which of these arcs lead to a state that is not
+    // accessible yet.
+    Renumbering arc_renumbering(c_, arcs_shape.NumElements());
+
+    const int32_t *arcs_row_ids1_data = arcs_shape.RowIds(1).Data(),
+                  *arcs_row_ids2_data = arcs_shape.RowIds(2).Data(),
+                  *arcs_row_splits2_data = arcs_shape.RowSplits(2).Data(),
+                  *fsas_row_splits1_data = fsas_.RowSplits(1).Data(),
+                  *dest_states_data = dest_states_.values.Data();
+    char *keep_arc_data = arc_renumbering.Keep().Data();
+    int32_t *next_iter_states_data = next_iter_states.Data();
+    K2_EVAL(
+        c_, arcs_shape.NumElements(), lambda_set_arc_renumbering,
+        (int32_t arcs_idx012)->void {
+          // note: the prefix `arcs_` means it is an idxXXX w.r.t. `arcs_shape`.
+          // the prefix `fsas_` means the variable is an idxXXX w.r.t. `fsas_`.
+          int32_t arcs_idx01 = arcs_row_ids2_data[arcs_idx012],
+                  arcs_idx01x = arcs_row_splits2_data[arcs_idx01],
+                  arcs_idx2 = arcs_idx012 - arcs_idx01x,
+                  fsas_idx01 = states_data[arcs_idx01],  // a state index
+                  fsas_idx01x = fsas_row_splits2_data[fsas_idx01],
+                  fsas_idx012 = fsas_idx01x + arcs_idx2,
+                  fsas_dest_state_idx01 = dest_states_data[fsas_idx012];
+          // if this arc is a self-loop, just ignore this arc as we won't
+          // processe the dest_state (current state) again
+          if (fsas_dest_state_idx01 == fsas_idx01 ||
+              accessible_data[fsas_dest_state_idx01]) {
+            keep_arc_data[arcs_idx012] = 0;
+            return;
+          }
+          keep_arc_data[arcs_idx012] = 1;
+          next_iter_states_data[arcs_idx012] = fsas_dest_state_idx01;
```

Comment:
Just to make you know, there maybe a chance to optimize the code so that we can save one kernel:
1. declare `new_states_row_ids` as `Array1<int32_t> new_states_row_ids(c_, arcs_shape.NumElements());`  and set its values in this kernel (i.e. `lambda_set_arc_renumbering`).
2. replace this line `Array1<int32_t> new_states = next_iter_states[new2old_map]` (this is also a kenerl) with one kernel, get `new_states` from `next_iter_stats` and `ans_row_ids` from `new_states_row_ids` in that kernel, so we don't need kernel `lambda_set_row_ids` anymore.

There maybe some empty fsas in the `ans` but I think it's fine. 


Noted you can also allocate GPU memory once for `next_iter_states_data` and `new_stats_row_ids` as below to get more perfomace gain (the gain may be small if our GPU memory cache machnism works here):
```
int32_t num_arcs = arc_shape.NumElements();
Array1<int32_t> temp(c_,  2*num_arcs);
Array1<int32_t> new_states_row_ids = temp.Arange(0, num_arcs);
Array1<int32_t> next_iter_states = temp.Arange(num_arcs, 2*num_arcs);
```
You can do the same thing for `new_states` and `ans_row_ids`.

Reply:
I see, will optimize it.